src/front/glsl/constants.rs

```diff
@@ -201,6 +209,47 @@ impl<'a> ConstantSolver<'a> {
                         let inner = ConstantInner::Scalar { width, value };
                         Ok(self.register_constant(inner, span))
                     }
+                    crate::MathFunction::Clamp => {
+                        let (value, width) = match (const0, const1.unwrap(), const2.unwrap()) {
+                            (
+                                &ConstantInner::Scalar {
+                                    width,
+                                    value: value0,
+                                },
+                                &ConstantInner::Scalar { value: value1, .. },
+                                &ConstantInner::Scalar { value: value2, .. },
+                            ) => (
+                                match (value0, value1, value2) {
+                                    (
+                                        ScalarValue::Sint(a),
+                                        ScalarValue::Sint(b),
+                                        ScalarValue::Sint(c),
+                                    ) => ScalarValue::Sint(a.max(b).min(c)),
+                                    (
+                                        ScalarValue::Uint(a),
+                                        ScalarValue::Uint(b),
+                                        ScalarValue::Uint(c),
+                                    ) => ScalarValue::Uint(a.max(b).min(c)),
+                                    (
+                                        ScalarValue::Float(a),
+                                        ScalarValue::Float(b),
+                                        ScalarValue::Float(c),
+                                    ) => ScalarValue::Float(a.max(b).min(c)),
```

Comment:
We need to figure out how the `NaN`s work with this. The glsl spec says
> min returns the minimum of the two parameters. It returns y if y is less than x, otherwise it returns x. 

while the rust standard library has

> If one of the arguments is NaN, then the other argument is returned.

I couldn't find how glsl defines the float ordering with `NaN`s, but standard `IEEE 754-2008` floats don't have a stable ordering between `NaN`s and any other number.

Reply:
I'm so sorry for the delay. I've been meaning to get back to this, but I got a little too busy... But I am meaning to resume this unresolved issue, so I was wondering if there was any update on how the `NaN`s work with clamp. In the meantime, I'll also try digging into what I can find :)

Reply:
I did found out how it worked, for `min` thanks to it's definition it's behavior is the same as rust. but for `max` this isn't the case so you'll need the following function
```rust
/// Helper function to implement the glsl `max` function for floats
///
/// While rust does provide a `f32::max` method it has a different behavior than the
/// glsl `max` for `NaN`s, in rust if any of the arguments is a `NaN` then the other
/// is returned.
///
/// This leads to different results in the following example
/// ```
/// max(NaN, 1.0)
/// ```
///
/// Rust will return 1.0 while glsl should return NaN
fn glsl_float_max(x: f32, y: f32) -> f32 {
    if x < y {
        y
    } else {
        x
    }
}
```

Reply:
Wait sorry used some old notes, `min` is not equal between the two. you also need this function
```rust
/// Helper function to implement the glsl `min` function for floats
///
/// While rust does provide a `f32::min` method it has a different behavior than the
/// glsl `min` for `NaN`s, in rust if any of the arguments is a `NaN` then the other
/// is returned.
///
/// This leads to different results in the following example
/// ```
/// min(NaN, 1.0)
/// ```
///
/// Rust will return 1.0 while glsl should return NaN
fn glsl_float_min(x: f32, y: f32) -> f32 {
    if y < x {
        y
    } else {
        x
    }
}
```

Reply:
Thanks so much for this! I'm almost ready to push the new change, but I am wondering in which file I should be writing that function? `glsl/functions.rs`?

Reply:
No it should be added in the same file since it concerns constant expression functions and not runtime functions

Reply:
Okay, thank you very much!

Reply:
I just pushed a new commit. Let me know if it works :) And thank you very much for your assist!
One thing maybe I should have gotten clarity on is that I changed `f32` to `f64` in the `glsl_float_min()` function, since `ScalarValue::Float` takes `f64`.

Reply:
You forgot to also use the `max` modified variant

Reply:
Hopefully fixed it :)

Reply:
Sorry didn't notice at first but you need to use `f64`s, when I posted the code I forgot that naga uses `f64`s

Reply:
Oh, no worries! That's what I was thinking too, but forgot to ask for clarity on it.